cf-custom-resources/lib/wkld-cert-validator.js

```diff
@@ -0,0 +1,881 @@
+      .listResourceRecordSets({
+        HostedZoneId: hostedZoneID,
+        MaxItems: "1",
+        StartRecordName: alias,
+        StartRecordType: "A",
+      })
+      .promise()
+      .then(({ ResourceRecordSets: recordSet }) => {
+        if (!targetRecordExists(alias, recordSet)) {
+          return;
+        }
+        if (recordSet[0].Type !== "A") {
+          return;
+        }
+        let aliasTarget = recordSet[0].AliasTarget;
+        if (
+          aliasTarget &&
+          loadBalancerDNS &&
+          aliasTarget.DNSName.toLowerCase() ===
+            `${loadBalancerDNS.toLowerCase()}.`
+        ) {
+          return; // The record is an alias record and is in use by myself, hence valid.
+        }
+        if (aliasTarget) {
+          throw new Error(
+            `Alias ${alias} is already in use by ${aliasTarget.DNSName}. This could be another load balancer of a different service.`
+          );
+        }
+        throw new Error(`Alias ${alias} is already in use`);
+      });
+    promises.push(promise);
+  }
+  await Promise.all(promises);
+}
+
+/**
+ * Requests a public certificate from AWS Certificate Manager, using DNS validation.
+ *
+ * @param {Object} requestCertificateInput is the input to requestCertificate, containing the alias and idempotencyToken.
+ * @return {String} The ARN of the requested certificate.
+ */
+async function requestCertificate({ aliases, idempotencyToken }) {
+  const { CertificateArn } = await clients
+    .acm()
+    .requestCertificate({
+      DomainName: certificateDomain,
+      IdempotencyToken: idempotencyToken,
+      SubjectAlternativeNames: aliases.size === 0 ? null : [...aliases],
+      Tags: [
+        {
+          Key: "copilot-application",
+          Value: appName,
+        },
+        {
+          Key: "copilot-environment",
+          Value: envName,
+        },
+        {
+          Key: "copilot-service",
+          Value: serviceName,
+        },
+      ],
+      ValidationMethod: "DNS",
+    })
+    .promise();
+  return CertificateArn;
+}
+
+/**
+ * Wait until the validation options are ready
+ *
+ * @param certificateARN
+ * @param {Set<String>} aliases for the service.
+ */
+async function waitForValidationOptionsToBeReady(certificateARN, aliases) {
+  let expectedCount = aliases.size + 1; // Expect one validation option for each alias and the cert domain.
+
+  let attempt; // TODO: This wait loops could be further abstracted.
+  for (attempt = 0; attempt < ATTEMPTS_VALIDATION_OPTIONS_READY; attempt++) {
+    let readyCount = 0;
+    const { Certificate } = await clients
+      .acm()
+      .describeCertificate({
+        CertificateArn: certificateARN,
+      })
+      .promise();
+    const options = Certificate.DomainValidationOptions || [];
+    options.forEach((option) => {
+      if (
+        option.ResourceRecord &&
+        (aliases.has(option.DomainName) ||
+          option.DomainName.toLowerCase() === certificateDomain.toLowerCase())
```

Comment:
same nit about it being easier to read on the same line
```suggestion
        (aliases.has(option.DomainName) ||  option.DomainName.toLowerCase() === certificateDomain.toLowerCase())
```